test/method.jl

```diff
@@ -234,4 +234,46 @@ end
             only_method(OneParamStruct{Float32}, Tuple{Float32, Bool})
         )
     end
+
+    @testset "signature(type_tuple)" begin
+        # our tests here are much less comprehensive than for `signature(::Method)`
+        # but that is OK, as most of the code is shared between the two
+
+        @test signature(Tuple{typeof(+), Float32, Float32}) == Dict(
+            # Notice the type of the function object is actually interpolated in to the Expr
+            # This is useful because it bypasses julia's pretection for overloading things
+            # which Nabla (and probably others generating overloads) depends upon
+            :name => :(op::$(typeof(+))),
+            :args => Expr[:(x1::Float32), :(x2::Float32)],
+        )
+
+        @test signature(Tuple{typeof(+), Array}) == Dict(
+            :name => :(op::$(typeof(+))),
+            :args => Expr[:(x1::(Array{T, N} where {T, N}))],
+        )
+
+        @test signature(Tuple{typeof(+), Vector{T}, Matrix{T}} where T<:Real) == Dict(
+            :name => :(op::$(typeof(+))),
+            :args => Expr[:(x1::Array{T, 1}), :(x2::Array{T, 2})],
+            :whereparams => Any[:(T <: Real)],
+        )
+
+        @testset "hygienic_unionalls" begin
+            no_hygiene = signature(Tuple{typeof(+),T,Array} where T)
+            @test no_hygiene == Dict(
+                :name => :(op::$(typeof(+))),
+                :args => Expr[:(x1::T), :(x2::(Array{T, N} where {T, N}))],
+                :whereparams => Any[:T],
+            )
+            hygiene = signature(Tuple{typeof(+),T,Array} where T; hygienic_unionalls=true)
+            @test no_hygiene[:name] == hygiene[:name]
+            @test length(no_hygiene[:args]) == 2
+            @test no_hygiene[:args][2] == hygiene[:args][2]
+            
+            @test length(no_hygiene[:whereparams]) == 1
+            @test no_hygiene[:whereparams] != hygiene[:whereparams]
```

Comment:
why is this different? I guess I don't really understand the hygienic part of renaming unionall. I thought that just the names are different, but they are the same object, similarly to e.g.
```julia
julia> u = Array{T, N} where {T<:Number, N}

julia> v = Array{T, M} where {T<:Number, M}

julia> u == v
true
```

Reply:
Those are `UnionAll`s, I assume equality for unionalls is defines as something like `u <: v && v <: u`
But the whereparams are the `Expr` and `Symbol`s that generate them and those are not equal.

The effect of the evaluttating `combinedef(signature(sig; hyginic_unionalls=false)` 
vs ``combinedef(signature(sig; hyginic_unionalls=true)`
should be identical.
But it isn't due to https://github.com/JuliaLang/julia/issues/39876